includes/create-theme/theme-patterns.php

```diff
@@ -47,4 +68,91 @@ public static function create_pattern_link( $attributes ) {
+
+	public static function replace_local_pattern_references( $pattern ) {
+		// List all template and pattern files in the theme
+		$base_dir       = get_stylesheet_directory();
+		$patterns       = glob( $base_dir . DIRECTORY_SEPARATOR . 'patterns' . DIRECTORY_SEPARATOR . '*.php' );
+		$templates      = glob( $base_dir . DIRECTORY_SEPARATOR . 'templates' . DIRECTORY_SEPARATOR . '*.html' );
+		$template_parts = glob( $base_dir . DIRECTORY_SEPARATOR . 'template-parts' . DIRECTORY_SEPARATOR . '*.html' );
+
+		// Replace references to the local patterns in the theme
+		foreach ( array_merge( $patterns, $templates, $template_parts ) as $file ) {
+			$file_content = file_get_contents( $file );
+			$file_content = str_replace( 'wp:block {"ref":' . $pattern->id . '}', 'wp:pattern {"slug":"' . $pattern->slug . '"}', $file_content );
+			file_put_contents( $file, $file_content );
+		}
+	}
+
+	public static function prepare_pattern_for_export( $pattern, $options = null ) {
+		if ( ! $options ) {
+			$options = array(
+				'localizeText'   => false,
+				'removeNavRefs'  => true,
+				'localizeImages' => true,
+			);
+		}
+
+		$pattern = CBT_Theme_Templates::eliminate_environment_specific_content( $pattern, $options );
+
+		if ( array_key_exists( 'localizeText', $options ) && $options['localizeText'] ) {
+			$pattern = CBT_Theme_Templates::escape_text_in_template( $pattern );
+		}
+
+		if ( array_key_exists( 'localizeImages', $options ) && $options['localizeImages'] ) {
+			$pattern = CBT_Theme_Media::make_template_images_local( $pattern );
+
+			// Write the media assets if there are any
+			if ( $pattern->media ) {
+				CBT_Theme_Media::add_media_to_local( $pattern->media );
+			}
+		}
+
+		return $pattern;
+	}
+
+	/**
+	 * Copy the local patterns as well as any media to the theme filesystem.
+	 */
+	public static function add_patterns_to_theme( $options = null ) {
+		$base_dir     = get_stylesheet_directory();
+		$patterns_dir = $base_dir . DIRECTORY_SEPARATOR . 'patterns';
+
+		$pattern_query = new WP_Query(
+			array(
+				'post_type'      => 'wp_block',
+				'posts_per_page' => -1,
+			)
+		);
+
+		if ( $pattern_query->have_posts() ) {
+			// If there is no patterns folder, create it.
+			if ( ! is_dir( $patterns_dir ) ) {
+				wp_mkdir_p( $patterns_dir );
+			}
+
+			foreach ( $pattern_query->posts as $pattern ) {
+				$pattern      = self::pattern_from_wp_block( $pattern );
+				$pattern      = self::prepare_pattern_for_export( $pattern, $options );
+				$pattern_file = $patterns_dir . 'pattern-' . $pattern->name . '.php';
+				file_put_contents(
+					$patterns_dir . DIRECTORY_SEPARATOR . 'pattern-' . $pattern->name . '.php',
+					$pattern->content
+				);
+
+				self::replace_local_pattern_references( $pattern );
+
+				/**
+				 * if the pattern is not synced then remove if from
+				 * database so patterns are loaded from the theme
+				 * and not the database
+				 */
+
+				if ( stripos( $pattern->content, 'pattern-overrides' ) !== true ) {
```

Comment:
This looks like a neat way to check if the pattern is not synced. I already mentioned [here](https://github.com/WordPress/create-block-theme/pull/675#issuecomment-2260911432), but I also found `wp_pattern_sync_status` in postmeta that we could use as well.